---
title: "GSE118545_P6"
author: "Cass Li"
date: "04/08/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<style>
div.teal pre { background-color:#fff44f; }
div.teal pre.r { background-color:#b2d8d8; }
</style>

<div class = "teal">
```{r library_loading, message = FALSE, warning = FALSE}
# load libraries
library(Seurat)
library(ggplot2)
library(homologene)
library(clustree)
library(dplyr)
library(topGO)
library(org.Mm.eg.db)
library(scales)
library(GSEABase)
library(AUCell)
library(plotly)
library(scran)
```
</div>

The authors applied sNucDrop-Seq, a droplet microfluidics-based massively parallel snRNA-Seq method, to investigate the transcriptional landscape of postnatal maturing mouse hearts in both healthy or disease state. By profiling the transcriptome of nearly 20,000 nuclei, they identified major and rare cardiac cell types and revealed significant heterogeneity of cardiomyocytes, fibroblasts and endothelial cells in the postnatal developing heart. I will use 6 datasets from their study, 3 biological replicates for each of the two ages, P6 and P10.

### Step 1 Load datasets for P6.

<div class = "teal">
```{r cache = TRUE}
P6_rep1_data <- read.delim("~/Documents/Project_scRNA_seq_meta_analysis/MetaAnalysis/RawFiles/GSE118545/GSM3332251_Ctrl-P6-1_gene_exonic.intronic_tagged.dge.txt.gz")
P6_rep1_data_1 <- P6_rep1_data[, -1]
rownames(P6_rep1_data_1) <- P6_rep1_data[, 1]
P6_rep1 <- CreateSeuratObject(counts = P6_rep1_data_1, project = "P6_rep1", min.cells = 3, min.features = 200)
P6_rep1[["percent.mt"]] <- PercentageFeatureSet(P6_rep1, pattern = "^mt-")
# fix the metadata
P6_rep1@meta.data$age <- "P6"
P6_rep1@meta.data$species <- "Mouse"
P6_rep1@meta.data$rep_no <- 1
# Visualise QC metrics
png(filename = "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/P6rep1QC.png", width = 600, height = 400)
VlnPlot(P6_rep1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0)
dev.off()
# Filtering of each dataset is based on the same cut off
P6_rep1 <- subset(P6_rep1, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 20)
# Calculate cell cycle scores - these are human genes and can't be used directly. Use homologene package to solve the problem
s.genes <- human2mouse(cc.genes$s.genes) # return a dataframe
g2m.genes <- human2mouse(cc.genes$g2m.genes) # return a dataframe
P6_rep1 <- NormalizeData(P6_rep1) # normalise before cell cycle score
P6_rep1 <- CellCycleScoring(P6_rep1, s.features = s.genes$mouseGene, g2m.features = g2m.genes$mouseGene[-49], set.ident = TRUE) # need to remove Nek2 from the g2m gene list as it's not in any of the datasets. Error still occurs when running it and try the solution in the issue #1227 solution by running a normalisation step first and it works!!
# Use scTransform to normalise datasets
P6_rep1 <- SCTransform(P6_rep1, vars.to.regress = c("percent.mt", "S.Score", "G2M.Score")) # default assay is now set to SCT

P6_rep2_data <- read.delim("~/Documents/Project_scRNA_seq_meta_analysis/MetaAnalysis/RawFiles/GSE118545/GSM3332252_Ctrl-P6-2_gene_exonic.intronic_tagged.dge.txt.gz")
P6_rep2_data_1 <- P6_rep2_data[, -1]
rownames(P6_rep2_data_1) <- P6_rep2_data[, 1]
P6_rep2 <- CreateSeuratObject(counts = P6_rep2_data_1, project = "P6_rep2", min.cells = 3, min.features = 200)
P6_rep2[["percent.mt"]] <- PercentageFeatureSet(P6_rep2, pattern = "^mt-")
# fix the metadata
P6_rep2@meta.data$age <- "P6"
P6_rep2@meta.data$species <- "Mouse"
P6_rep2@meta.data$rep_no <- 2
# Visualise QC metrics
png(filename = "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/P6rep2QC.png", width = 600, height = 400)
VlnPlot(P6_rep2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0)
dev.off()
# Filtering of each dataset is based on the same cut off
P6_rep2 <- subset(P6_rep2, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 20)
# calculate cell cycle score
P6_rep2 <- NormalizeData(P6_rep2) # normalisation required before cell cycle scoring
P6_rep2 <- CellCycleScoring(P6_rep2, s.features = s.genes$mouseGene, g2m.features = g2m.genes$mouseGene[-49], set.ident = TRUE)
# Use scTransform to normalise datasets
P6_rep2 <- SCTransform(P6_rep2, vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"))

P6_rep3_data <- read.delim("~/Documents/Project_scRNA_seq_meta_analysis/MetaAnalysis/RawFiles/GSE118545/GSM3332253_Ctrl-P6-3_gene_exonic.intronic_tagged.dge.txt.gz")
P6_rep3_data_1 <- P6_rep3_data[, -1]
rownames(P6_rep3_data_1) <- P6_rep3_data[, 1]
P6_rep3 <- CreateSeuratObject(counts = P6_rep3_data_1, project = "P6_rep3", min.cells = 3, min.features = 200)
P6_rep3[["percent.mt"]] <- PercentageFeatureSet(P6_rep3, pattern = "^mt-")
# fix the metadata
P6_rep3@meta.data$age <- "P6"
P6_rep3@meta.data$species <- "Mouse"
P6_rep3@meta.data$rep_no <- 3
# Visualise QC metrics
png(filename = "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/P6rep3QC.png", width = 600, height = 400)
VlnPlot(P6_rep3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0)
dev.off()
# Filtering of each dataset is based on the same cut off
P6_rep3 <- subset(P6_rep3, subset = nFeature_RNA > 200 & nFeature_RNA < 4000 & percent.mt < 20)
# calculate cell cycle score
P6_rep3 <- NormalizeData(P6_rep3) # normalisation required before cell cycle scoring
P6_rep3 <- CellCycleScoring(P6_rep3, s.features = s.genes$mouseGene, g2m.features = g2m.genes$mouseGene[-49], set.ident = TRUE) # Nek2 removed
P6_rep3 <- SCTransform(P6_rep3, vars.to.regress = c("percent.mt", "S.Score", "G2M.Score"))

# combine the QC data
qc_rep1 <- P6_rep1@meta.data[, 1:5]
qc_rep2 <- P6_rep2@meta.data[, 1:5]
qc_rep3 <- P6_rep3@meta.data[, 1:5]
qc <- rbind(qc_rep1, qc_rep2, qc_rep3)

# use 'download as png' to save
plot_ly(qc, y = ~ nCount_RNA, color = ~ orig.ident, type = "box", colors = c("#383745","#A17724", "#9E9CC2")) %>% layout(showlegend = F, autosize = F, width = 500, height = 500, xaxis = list(tickfont = list(size = 15), titlefont = list(size = 15)), yaxis = list(title = "nCount_RNA", tickfont = list(size = 15), titlefont = list(size = 15)), legend = list(font = list(size = 15)))

plot_ly(qc, y = ~ nFeature_RNA, color = ~ orig.ident, type = "box", colors = c("#383745","#A17724", "#9E9CC2")) %>% layout(showlegend = F, autosize = F, width = 500, height = 500, xaxis = list(tickfont = list(size = 15), titlefont = list(size = 15)), yaxis = list(title = "nFeature_RNA", tickfont = list(size = 15), titlefont = list(size = 15)), legend = list(font = list(size = 15)))

plot_ly(qc, y = ~ percent.mt, color = ~ orig.ident, type = "box", colors = c("#383745","#A17724", "#9E9CC2")) %>% layout(showlegend = F, autosize = F, width = 500, height = 500, xaxis = list(tickfont = list(size = 15), titlefont = list(size = 15)), yaxis = list(title = "percent.mt", tickfont = list(size = 15), titlefont = list(size = 15)), legend = list(font = list(size = 15)))
```
</div>

### Step 2 Integrate 3 replicates for P6.

All three datasets are now SCTransformed and the default assay is set to "SCT".

<div class = "teal">
```{r integrate_P6_replicates, warnings = FALSE, message = FALSE}
options(future.globals.maxSize = 1000 * 1024^2)
# select features for integration
P6_features <- SelectIntegrationFeatures(object.list = c(P6_rep1, P6_rep2, P6_rep3), nfeatures = 3000)
# it worked
P6_list <- PrepSCTIntegration(object.list = c(P6_rep1, P6_rep2, P6_rep3), anchor.features = P6_features, verbose = FALSE) # error: exceeds the maximum allowed size of 500.00 MiB and adjust the limit to 1 GB in the previous line
# Identify the anchors - this step takes a while
P6_anchors <- FindIntegrationAnchors(object.list = P6_list, normalization.method = "SCT", anchor.features = P6_features, verbose = FALSE) # already renamed
# Integrate data
P6_integrated <- IntegrateData(anchorset = P6_anchors, normalization.method = "SCT", verbose = FALSE) # this step reported the existence of duplicated cell names
## check duplicated cell names - 1 cell between rep1 and rep2 and 1 cell between rep2 and rep3 and need to remove these two cells from the datasets and redo the integration but NOT NOW, in a hurry
intersect(rownames(P6_rep1@meta.data), rownames(P6_rep2@meta.data))  
## [1] "ATTATTAAATCC"
intersect(rownames(P6_rep2@meta.data), rownames(P6_rep3@meta.data))  
## [1] "AGATCCTTACCC"
intersect(rownames(P6_rep1@meta.data), rownames(P6_rep3@meta.data))  
## character(0)
```
</div>

### Step 3 Proceed with downstream analysis on the integrated dataset

<div class = "teal">
```{r downstream_analysis, message = FALSE, warning = FALSE}
P6_integrated <- RunPCA(P6_integrated, verbose = FALSE)
set.seed(1024)
P6_integrated <- RunUMAP(P6_integrated, dims = 1:30)
saveRDS(P6_integrated, file = "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated.rds")
png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_Integrated_UMAP_by_rep_no.png", res = 300, width = 2500, height = 2500)
DimPlot(P6_integrated, combine = FALSE, group.by = "rep_no", pt.size = 1, cols = c("#383745","#A17724", "#9E9CC2"))
dev.off()
png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_Integrated_UMAP_split_by_rep_no.png", res = 300, width = 7500, height = 2500)
DimPlot(P6_integrated, combine = FALSE, split.by = "rep_no", pt.size = 1)
dev.off()
```
</div>

### Step 4 FindNeighbors and FindClusters (on default assay which is "integrated")

<div class = "teal">
```{r clustering, message = FALSE, warning = FALSE}
P6_integrated <- FindNeighbors(P6_integrated, dims = 1:30)

for (i in 1:20) {
        P6_integrated <- FindClusters(P6_integrated, resolution = 0.1*i)
}
png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_clustree.png", res = 300, width = 4000, height = 3000)
clustree(P6_integrated) # cluster tree visualisation
dev.off()
png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_clustree_stability.png", res = 300, width = 4000, height = 3000)
clustree(P6_integrated, node_colour = "sc3_stability") # cluster tree visualisation
dev.off()
# It seems resolution 0.8 is a good choice and change the clusters to the res 0.8 one
P6_integrated <- FindClusters(P6_integrated, resolution = 0.8) # this resolution is difficult to determine and the package "clustree" is used to help with all resolution between 0.1 and 2 (with a step of 0.1) tested
P6_integrated@meta.data$seurat_clusters <- P6_integrated@meta.data$integrated_snn_res.0.8

png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_clustering_P6_integrated.png", res = 300, width = 2500, height = 2500)
DimPlot(P6_integrated, label = TRUE, label.size = 6) + NoLegend()
p <- DimPlot(P6_integrated, label = TRUE, label.size = 6) + NoLegend()
ggplotly(p, width = 500, height = 500)
dev.off()

png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_clustering_P6_integrated.png", res = 300, width = 7500, height = 2500)
DimPlot(P6_integrated, label = TRUE, label.size = 6, split.by = "rep_no") + NoLegend()
dev.off()

# Find cluster markers using scran
P6_sce <- as.SingleCellExperiment(P6_integrated)
P6_sce_markers <- findMarkers(P6_sce, groups = P6_sce$ident)
chosen <- "9" # choose cluster 9
interesting <- P6_sce_markers[[chosen]]
interesting[1:10, 1:4]
best.set <- interesting[interesting$Top <= 6, ]
# getMarkerEffects is in scran v1.16.0 but failed to install to R v3.6
getMarkerEffects <- function(x, prefix="logFC", strip=TRUE, remove.na.col=FALSE) {
    regex <- paste0("^", prefix, "\\.")
    i <- grep(regex, colnames(x))
    out <- as.matrix(x[,i])

    if (strip) {
        colnames(out) <- sub(regex, "", colnames(out))
    }
    if (remove.na.col) {
        out <- out[,!colAnyNAs(out),drop=FALSE]
    }

    out
}

logFCs <- getMarkerEffects(best.set)
library(pheatmap)
pheatmap(logFCs, bbreaks = seq(-5, 5, length.out = 101))

# use wilcox test to find cluster specific markers
wmw_P6_sce_markers <- findMarkers(P6_sce, test = "wilcox", direction = "up", groups = P6_sce$ident)
wmw_interesting <- wmw_P6_sce_markers[[chosen]]
wmw_interesting[1:10, 1:4]
best.set <- wmw_interesting[wmw_interesting$Top <= 5, ]
AUCs <- getMarkerEffects(best.set, prefix = "AUC")

pheatmap(AUCs, breaks = seq(0, 1, length.out = 21), color = viridis::viridis(21))

# Find cluster markers using the clustering information in this assay
P6_integrated_markers <- FindAllMarkers(P6_integrated, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
P6_integrated_markers_top10 <- P6_integrated_markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
write.csv(P6_integrated_markers_top10, file = "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated_cluster_marker_top10.csv")
# Use topGO to annotate the cluster markers
# need to install BiocManager::install("org.Mm.eg.db") if not already
## cluster 0 is done first - build a loop to get all clusters' GO
length <- length(unique(P6_integrated@meta.data$seurat_clusters)) # number of loops
for (i in 0:(length - 1)) { # starting from 0
        cluster <- subset(P6_integrated, idents = as.character(i))
        expr <- cluster@assays$integrated@data
        # Select genes that are expressed > 0 in at least 75% of cells (somewhat arbitrary definition and too high in my case)
        n.gt.0 <- apply(expr, 1, function(x)length(which(x > 0)))
        expressed.genes <- rownames(expr)[which(n.gt.0/ncol(expr) >= 0.60)]
        all.genes <- rownames(expr)
        # define geneList as 1 if gene is in expressed.genes, 0 otherwise
        geneList <- ifelse(all.genes %in% expressed.genes, 1, 0)
        names(geneList) <- all.genes
        # Create topGOdata object
        GOdata <- new("topGOdata",
                      ontology = "BP", # use biological process ontology
                      allGenes = geneList,
                      geneSelectionFun = function(x)(x == 1),
                      annot = annFUN.org, mapping = "org.Mm.eg.db", ID = "symbol")
        # Test for enrichment using Fisher's Exact Test
        resultFisher <- runTest(GOdata, algorithm = "elim", statistic = "fisher")
        GO_table <- GenTable(GOdata, Fisher = resultFisher, topNodes = 20, numChar = 60)
        # Save GO table
        write.csv(GO_table, file = paste0("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated_cluster_", i,"GO_term_table", ".csv"))
        # Generate GO term bar graph
        Go_table_touse <- GO_table[1:5, ] # choose the top 5 GO terms
        Go_table_touse$NewTerm <- paste0(Go_table_touse$GO.ID, "_", Go_table_touse$Term)
        Go_table_touse$P_value <- -log10(as.numeric(Go_table_touse$Fisher))
        colour_touse <- hue_pal()(length)[i+1]
        png(paste0("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated_cluster_", i,"GO_term_table", ".png"), res = 300, width = 4000, height = 1000)
        print(ggplot(Go_table_touse, aes(x = NewTerm, y = P_value)) +
                geom_bar(stat="identity", fill = colour_touse, width = 0.6) + 
                coord_flip() +
                ggtitle(paste0("P6 Cluster ", i)) +
                ylab(expression("-log"[10]~ paste("(", italic(P), " value)"))) +
                theme_classic() + 
                theme(axis.line = element_line(colour = "black", size = 1, linetype = "solid"), axis.text.x = element_text(face = "bold", color = "black", size = 14), axis.text.y = element_text(face="bold", color = "black", size = 14), axis.title.y = element_blank(), plot.title = element_text(size = 14, face = "bold"), axis.title.x = element_text(color = "black", size = 14, face = "bold"))) # Remember to wrap the png call using print to get it plotted and saved!!!!!!!!
        dev.off()
}
```
</div>

### Step 5 Change default assay to "SCT"; normalize then generate FeaturePlots 

The reason of changing back to RNA assay is because the integrated assay only has 3000 genes and values from SCT assay represent the SCT normalized values for each dataset, prior to integration.

<div class = "teal">
```{r feature_plot, warning = FALSE, message = FALSE}
png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated_RNAassay_Pecam1_feature_plot.png", res = 300, width = 2500, height = 2500)
FeaturePlot(P6_integrated, features = c("Pecam1"), cols = c("lightgrey", "#003333"), pt.size = 1)
p <- FeaturePlot(P6_integrated, features = c("Pecam1"), cols = c("lightgrey", "#003333"), pt.size = 1)
ggplotly(p, width = 500, height = 500)
dev.off()
```
</div>

### Step 6 Extract Endothelial cells using AUCell

Since the scoring method is rank based, it is independent of the gene expression units and the normalisation procedure.

<div class = "teal">
```{r feature_plot, warning = FALSE, message = FALSE}
saveRDS(P6_integrated, "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated_object.rds") # save object 
# Load count data
exprMatrix <- P6_integrated@assays$integrated@data
exprMatrix <- as.matrix(exprMatrix) # convert sparse matrix to matrix
dim(exprMatrix)
# Set gene set
ECgenes <- c("Pecam1", "Itgb1", "Cd34", "Cd36", "Entpd1", "Cd44", "Cd47", "Icam1", "Itgb3", "Selplg", "Cd80", "Cd86", "Cd93", "Icam2", "Eng", "Vcam1", "Il1r1", "Thbd", "F3", "Ace", "Cdh5", "Mcam", "Bsg", "Cd151", "Cd160", "Pik3cd", "Procr", "Il13ra1", "Cd248", "Kdr", "Adam8", "Adam9", "Adam10", "Adam12", "Adam15", "Adam17", "Adam33", "Adamts13", "Adamts18", "Cxcl16", "Dcbld2", "Emcn", "Esam", "Itga4", "Klf4")
geneSets <- GeneSet(ECgenes, setName = "endothelial_cells")
geneSets
# Build gene expression rankings for each cell
detectCores()
cells_rankings <- AUCell_buildRankings(exprMatrix, nCores = 2, plotStats = TRUE)
cells_rankings
save(cells_rankings, file = "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_EC_rankings.RData")
# Calculate enrichment for the gene signatures (AUC)
cells_AUC <- AUCell_calcAUC(geneSets, cells_rankings) 
# add AUC score to the P6_integrated object
P6_integrated@meta.data$AUC_score <- t(getAUC(cells_AUC))
save(cells_AUC, file = "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_EC_AUC.RData")
# Determine the cells with the given gene signatures or active gene sets
set.seed(123)
cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist = TRUE, assign = TRUE)
cells_assignment$endothelial_cells$aucThr$thresholds

endoAssigned <- cells_assignment$endothelial_cells$assignment
length(endoAssigned)
# 1511 cells in 8866 cells

geneSetName <- rownames(cells_AUC)[grep("endothelial_cells", rownames(cells_AUC))]
png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated_RNAassay_AUCell_EC_hist.png", res = 300, width = 1500, height = 1500)
AUCell_plotHist(cells_AUC[geneSetName, ], aucThr = cells_assignment$endothelial_cells$aucThr$thresholds[2, 1])
abline(v = cells_assignment$endothelial_cells$aucThr$thresholds[2, 1])
dev.off()

# AUCscore as feature plot
p <- FeaturePlot(P6_integrated, features = "AUC_score", cols = viridis::viridis(21)) + NoLegend()
ggplotly(p, width = 500, height = 500)

png("~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_integrated_RNAassay_AUCell_EC.png", res = 300, width = 2750, height = 2500)
DimPlot(P6_integrated, cells.highlight = WhichCells(object = P6_integrated, expression = AUC_score > cells_assignment$endothelial_cells$aucThr$thresholds[2, 1])) + scale_color_manual(labels = c("non-EC", "EC"), values = c("grey", "red")) + labs(color = "Cell type")
p <- DimPlot(P6_integrated, cells.highlight = WhichCells(object = P6_integrated, expression = AUC_score > cells_assignment$endothelial_cells$aucThr$thresholds[2, 1]), sizes.highlight = NULL, cols.highlight = "red") + labs(title = "Endothelial Cell Assignment") + NoLegend() + theme(plot.title = element_text(hjust = 0.5))
ggplotly(p, width = 500, height = 500)
dev.off()

# Subset and store the EC data
P6_EC <- subset(P6_integrated, cells = names(which(getAUC(cells_AUC)["endothelial_cells",] > cells_assignment$endothelial_cells$aucThr$thresholds[2, 1])))
saveRDS(P6_EC, "~/Documents/Project_scRNA_seq_meta_analysis/P6_P10_Adult_Healthy_mouse_meta/GSE118545_P6_EC.rds")
```
</div>
